src/main/java/com/rednavis/vaadin/showcase/backend/service/password/PasswordServiceImpl.java

```diff
@@ -0,0 +1,285 @@
+  private static final int PBKDF2_INDEX = 4;
+
+  private final PasswordValidator passwordValidator;
+
+  @Inject
+  public PasswordServiceImpl(PasswordConfig passwordConfig) {
+    List<Rule> ruleList = new ArrayList<>();
+    ruleList.add(new LengthRule(passwordConfig.getMinLength(), passwordConfig.getMaxLength()));
+
+    if (passwordConfig.getRequireLeastNumberOfDigits() > 0) {
+      ruleList.add(new DigitCharacterRule(passwordConfig.getRequireLeastNumberOfDigits()));
+    }
+    if (passwordConfig.getRequireLeastNumberOfNotAlphanumerics() > 0) {
+      ruleList.add(new NonAlphanumericCharacterRule(passwordConfig.getRequireLeastNumberOfNotAlphanumerics()));
+    }
+    if (passwordConfig.getRequireLeastNumberOfLowercases() > 0) {
+      ruleList.add(new LowercaseCharacterRule(passwordConfig.getRequireLeastNumberOfLowercases()));
+    }
+    if (passwordConfig.getRequireLeastNumberOfUppercases() > 0) {
+      ruleList.add(new UppercaseCharacterRule(passwordConfig.getRequireLeastNumberOfUppercases()));
+    }
+    if (passwordConfig.getAllowRepeatCharactersMax() > 0) {
+      ruleList.add(new RepeatCharacterRegexRule(passwordConfig.getAllowRepeatCharactersMax()));
+    }
+    if (!passwordConfig.isWhitespaceAllowed()) {
+      ruleList.add(new WhitespaceRule());
+    }
+    if (!passwordConfig.isAllowAlphabeticalSequence()) {
+      ruleList.add(new AlphabeticalSequenceRule());
+    }
+    if (!passwordConfig.isAllowNumericalSequence()) {
+      ruleList.add(new NumericalSequenceRule());
+    }
+    if (!passwordConfig.isAllowQwertySequence()) {
+      ruleList.add(new QwertySequenceRule());
+    }
+    if (passwordConfig.getRequireLeastNumberOfAlphabeticals() > 0) {
+      ruleList.add(new AlphabeticalCharacterRule(passwordConfig.getRequireLeastNumberOfAlphabeticals()));
+    }
+
+    passwordValidator = new PasswordValidator(ruleList);
+  }
+
+  public static String createHash(String password) throws CannotPerformOperationException {
+    return createHash(password.toCharArray());
+  }
+
+  public static String createHash(char[] password) throws CannotPerformOperationException {
+    // Generate a random salt
+    SecureRandom random = new SecureRandom();
+    byte[] salt = new byte[SALT_BYTE_SIZE];
+    random.nextBytes(salt);
+
+    // Hash the password
+    byte[] hash = pbkdf2(password, salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
+    int hashSize = hash.length;
+
+    // format: algorithm:iterations:hashSize:salt:hash
+    String parts = "sha1:" +
+        PBKDF2_ITERATIONS +
+        ":" + hashSize +
+        ":" +
+        toBase64(salt) +
+        ":" +
+        toBase64(hash);
+    return parts;
+  }
+
+  public static boolean verifyPassword(String password, String correctHash)
+      throws CannotPerformOperationException, InvalidHashException {
+    return verifyPassword(password.toCharArray(), correctHash);
+  }
+
+  public static boolean verifyPassword(char[] password, String correctHash)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'verifyPassword(char, String)' has an NPath complexity of 256](https://app.codacy.com/gh/rednavis/vaadin-showcase/pullRequest?prid=4839326)